[RFC] autocmds: add WinClosed command

maurelio1234 opened the pull request:
Fixes #3338

Commit:
autocmds: add WinClosed command

Commit:
clint: fix errors

Commit:
help: Add WinClosed event to vim_diff.txt

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -1053,7 +1054,9 @@ WinLeave			Before leaving a window.  If the window to be
-
+							*WinClosed*
+WinClosed			Before closing a window.  Just after WinLeave
+				and BufWinLeave.
```
If it's before closing, it should be named `WinClosePre`.

justinmk commented on test/functional/autocmd/autocmd_spec.lua:
```diff
@@ -34,6 +34,23 @@ describe('autocmds:', function()
 
+  it(':close triggers WinClose event', function()
+    command('let g:triggered = 0')
+    command('autocmd WinClosed * :let g:triggered=1')
```
It should be `let g:triggered+=1`, so that we can verify that it didn't fire more than once.

maurelio1234 replied:
Fixed that, there was actually a bug in it ;)

justinmk commented:
Also need to test for "recursion", i.e. the case where `WinClosed` event _causes_ another window to close. Usually we just disallow this, e.g. by skipping `apply_autocmd` if a `static int recursive` function-scoped variable is `>0`.

> WinClosed ... Before closing a window.  Just after WinLeave and BufWinLeave.

If it's before closing, it should be named `WinClosePre`.

Is there a reason for doing it before closing (that implies it's possible the window _does not close_ for some reason)? Will users also want a "after closing" event?

Commit:
typo: event name changed

Commit:
autocmd: only fire once, just before freeing mem

* Add test to make sure event just fires once
* Update documentation to clarify when event is fired

Commit:
typo: fixe WinClose in comments

maurelio1234 commented:
> Also need to test for "recursion", i.e. the case where `WinClosed` event _causes_ another window to close. Usually we just disallow this, e.g. by skipping `apply_autocmd` if a `static int recursive` function-scoped variable is `>0`.
> 
> >

I will think about that

[ ] test for recursion

maurelio1234 commented:
> If it's before closing, it should be named `WinClosePre`.
> 
> Is there a reason for doing it before closing (that implies it's possible the window _does not close_ for some reason)? Will users also want a "after closing" event?

- I clarified the text in my last commit, it is called after closing just before freeing resources (just like `BufUnload`). At least in the way it is coded now it is called when we know it is going to be closed. 

- I do not see the point of having a Pre/Post pair of events, looking at the current list of events, Pre/Post pairs seem to be useful when you are performing some operation to a somehow persistent resource (a file a register, a set of windows), so that you can change the resource before performing the operation.

- I am still not sure if `WinClosed` is a good name for the event :) `WinClose` seems more in line with how other events are named. What do you think about that

Commit:
WinClosed: triggered when on a different buffer

Commit:
test: WinClosed events are not called recursivelly

Commit:
refacto: use eq instead of assert.same

maurelio1234 commented:
Hi @justinmk,

I just added a test for the recursive case (https://github.com/neovim/neovim/pull/9066/files#diff-bdc18dc18b088a1cc092d1fce6face23R65) and the second level event is not fired.

Could you check if I understood your remark about recursive events correctly?

Commit:
test: WinClosed are not recursive in the same window

Commit:
clint: fix too long lines

Commit:
clint: fix indent and whitespace

Commit:
test: using funcs.bufwinnr instead of eval

It seems to break ci tests, even though they pass on my system.
Most of other tests call bufwinnr directly instead of eval'ing it.

Commit:
lint: fix whitespaces

Commit:
lint: fix whitespace

Commit:
win_close: fail if called recursively

maurelio1234 commented:
The remaining failed builds seem to be unrelated to my PR

Commit:
WinClosed: avoid recursive calls in another tab

justinmk commented:
This failure is releated:

```
[ RUN      ] autocmds: :WinClosed events are not recursive in different tab: FAIL
...d/neovim/neovim/test/functional/autocmd/autocmd_spec.lua:104: Expected objects to be the same.
Passed in:
(nil)
Expected:
(number) -1

stack traceback:
	...d/neovim/neovim/test/functional/autocmd/autocmd_spec.lua:104: in function <...d/neovim/neovim/test/functional/autocmd/autocmd_spec.lua:92>

==================== File /home/travis/build/neovim/neovim/build/log/ubsan.13158 ====================
= ../src/nvim/highlight.c:150:12: runtime error: member access within null pointer of type 'win_T' (aka 'struct window_S')
=     #0 0xe056c8 in update_window_hl /home/travis/build/neovim/neovim/build/../src/nvim/highlight.c:150:12
=     #1 0x147706b in update_screen /home/travis/build/neovim/neovim/build/../src/nvim/screen.c:391:9
=     #2 0x64383e in nvim_command /home/travis/build/neovim/neovim/build/../src/nvim/api/vim.c:60:3
=     #3 0x5831f8 in handle_nvim_command /home/travis/build/neovim/neovim/build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:1968:3
=     #4 0x10b1b27 in on_request_event /home/travis/build/neovim/neovim/build/../src/nvim/msgpack_rpc/channel.c:365:19
=     #5 0xa7bba7 in multiqueue_process_events /home/travis/build/neovim/neovim/build/../src/nvim/event/multiqueue.c:150:7
=     #6 0x116f29b in nv_event /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:8019:3
=     #7 0x111d180 in normal_execute /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:1137:3
=     #8 0x17970a9 in state_enter /home/travis/build/neovim/neovim/build/../src/nvim/state.c:73:26
=     #9 0x10d7dc4 in normal_enter /home/travis/build/neovim/neovim/build/../src/nvim/normal.c:467:3
=     #10 0xebe84f in main /home/travis/build/neovim/neovim/build/../src/nvim/main.c:595:3
=     #11 0x7f71329cef44 in __libc_start_main /build/eglibc-ripdx6/eglibc-2.19/csu/libc-start.c:287
=     #12 0x44e69b in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x44e69b)
= 
= SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/nvim/highlight.c:150:12 in 
=====================================================================================================
test/helpers.lua:149: assertion failed!

stack traceback:
	test/helpers.lua:149: in function 'check_logs'
	test/functional/helpers.lua:756: in function <test/functional/helpers.lua:752>
```

justinmk commented:
Likely that failure is because recursion isn't prevented. Look at `do_autocmd_dirchanged` for the typical way to prevent recursion. If `recursive` is true _do not_ trigger the WinClose event. There is no reasonable need for a `WinClose` event which deletes some other window to cause _another_ `WinClose` event. We will simply document this: at `:help WinClose` add a line that says `Recursion is ignored.`, like `:help DirChanged`.

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -1053,7 +1054,9 @@ WinLeave			Before leaving a window.  If the window to be
-
+							*WinClosed*
+WinClosed			After closing a window.  Just before freeing
+				window's resources.
```
Recursion is ignored.

Users don't care about "resources" and it's just confusing to mention that, it is never mentioned anywhere else in the help docs.

justinmk commented on src/nvim/auevents.lua:
```diff
@@ -101,6 +101,7 @@ return {
     'WinNew',                 -- when entering a new window
     'WinEnter',               -- after entering a window
     'WinLeave',               -- before leaving a window
+    'WinClosed',              -- after closing a window
```
should be in alphabetical order. I see that `ChanOpen` and `WinNew` are out of order, will fix those.

justinmk commented on src/nvim/window.c:
```diff
@@ -1928,9 +1928,11 @@ int win_close(win_T *win, bool free_buf)
     win->w_closing = true;
-    apply_autocmds(EVENT_WINLEAVE, NULL, NULL, FALSE, curbuf);
-    if (!win_valid(win))
+    apply_autocmds(EVENT_WINLEAVE, NULL, NULL, false, curbuf);
```
should be skipped if `recursive` (see other comments).

```c
if (!recursive) {
    apply_autocmds(EVENT_WINLEAVE, NULL, NULL, false, curbuf);
}
```

maurelio1234 replied:
Hi,

I do not see `recursive`  used anywhere in `window.c` nor in `buffer.c`, it is not even available from the current includes... where is it declared?

The code as it is now, tries to detect loops the way loops in other events are detected (checking if the window is still valid).

justinmk commented on src/nvim/window.c:
```diff
@@ -1939,6 +1941,16 @@ int win_close(win_T *win, bool free_buf)
+
+    // If autocommands closed the window, there's nothing else to do
+    if (!win_valid(win)) {
+      return FAIL;
```
Can you explain this? How could autocommands close the window during a `WinClosed` event for that very window?

maurelio1234 replied:
Yes, I'm trying to avoid recursion. But as I said before, I do not see where/how is `recursive` defined.

justinmk commented on src/nvim/window.c:
```diff
@@ -1939,6 +1941,16 @@ int win_close(win_T *win, bool free_buf)
   }
 
+  // fire WinClosed event just before freeing memory associated with window
+  if (has_event(EVENT_WINCLOSED)) {
```
should also check `!recursive` (see other comments).

should be skipped if `recursive` (see other comments).

```c
if (!recursive && has_event(EVENT_WINCLOSED)) {
  ...
```

maurelio1234 replied:
I said before, I do not see where/how is `recursive` defined.

justinmk commented on src/nvim/window.c:
```diff
@@ -2135,6 +2149,17 @@ void win_close_othertab(win_T *win, int free_buf, tabpage_T *tp)
+                   win->w_buffer->b_fname, false, win->w_buffer);
+
+    // If autocommands closed the window, there's nothing else to do
+    if (!win_valid_any_tab(win)) {
```
same comments as for the `win_close` case.

maurelio1234 replied:
The same comment as before

justinmk commented on test/functional/autocmd/autocmd_spec.lua:
```diff
@@ -34,6 +36,76 @@ describe('autocmds:', function()
+  it(':bdelete triggers WinClosed event', function()
+    command('let g:triggered = 0')
+    command('autocmd WinClosed <buffer> :let g:triggered+=1')
+    local first_buffer = eval('bufnr(\'%\')')
```
can avoid escaping by using double quotes (interchangeable with single-quotes, in Lua).

    local first_buffer = eval("bufnr('%')")

justinmk commented on test/functional/autocmd/autocmd_spec.lua:
```diff
@@ -34,6 +36,76 @@ describe('autocmds:', function()
+    eq(1, eval('g:triggered'))
+  end)
+
+  it(':WinClosed events are not recursive in different tab', function()
```
`WinClosed` is not a command, so no colon prefix :)

justinmk commented:
This is looking pretty good, just need to figure out these last questions :)

Commit:
comments: remove :

WinClosed is not a command

Commit:
style: avoid escaping

Commit:
doc: remove confusing wording

Commit:
doc: events in alphabetical order

maurelio1234 commented:
> This failure is related (or it exposes an existing bug):

Yes, it was introduced by the commits just after that comment... I saw I was missing some cases. Still working on them.

maurelio1234 commented:
> Likely that failure is because recursion isn't prevented. Look at `do_autocmd_dirchanged` for the typical way to prevent recursion. If `recursive` is true _do not_ trigger the WinClose event. There is no reasonable need for a `WinClose` event which closes another window to cause _another_ `WinClose` event. We will simply document this: at `:help WinClose` add a line that says `Recursion is ignored.`, like `:help DirChanged`.

I got what you mean now! I will test it :)

maurelio1234 commented:
I did some tests... but I'm not sure about which kinds of recursion you want to forbid here.
The problem is, there are two code paths for closing a window, either it is on the current tab or on another tab, what if closing the current window triggers a window closing on another tab, which triggers a the first window closing...

is it forbidden?

Commit:
WinClosed: static bool for recursive check

justinmk commented:
> what if closing the current window triggers a window closing on another tab, which triggers a the first window closing...

Then only the first WinClose is triggered. No recursion.

maurelio1234 committed:
recursive: check in do_autocmd_winclosed

maurelio1234 commented:
Hello, I just pushed a first take on that. The only missing point for me is the role of the field `w_closing` in windows... there's already a protection against a window causing itself to close recursively. I guess this field is not necessary anymore.

maurelio1234 committed:
Merge remote-tracking branch 'upstream/master' into issue_3338

justinmk commented:
still a ASAN failure. search the travis log for "AddressSanitizer" .

maurelio1234 committed:
asan: isolate variable use code

It looks like ASAN errors are not related to the new autocmd but to the
bdelete ex cmd that seems to break related to the deleted buffer.

maurelio1234 committed:
aucmd: moving calls to before starting releasing buffers

To make sure it won't try to use already freed resources on an autocmd.

maurelio1234 committed:
Revert "asan: isolate variable use code"

This reverts commit bebfd6b1c311cde39f0569285ced8c4381e9e813.

maurelio1234 committed:
asan: if abort if window was closed by autocmd

maurelio1234 committed:
window: win_valid_any_tab not win_valid

We may be on a different tab after autocmd runs

maurelio1234 committed:
lint: remove unused variable

maurelio1234 commented:
Hello the ASAN error should be fixed now. 
Some builds seem to be red on a missing tool (texi2pdf)

justinmk commented:
continued at #11331 (you will still get author credit)
